[DOCS][ESQL] Cleanup and cross-reference LOOKUP JOIN reference and landing pages

leemthompo opened the pull request:
Contributes to https://github.com/elastic/elasticsearch/issues/126897

## URL Previews

- [Command reference](https://docs-v3-preview.elastic.dev/elastic/elasticsearch/pull/127215/reference/query-languages/esql/commands/processing-commands#esql-lookup-join)
- [Landing page](https://docs-v3-preview.elastic.dev/elastic/elasticsearch/pull/127215/reference/query-languages/esql/esql-lookup-join)

## Summary 

Main goal is to ensure both pages have clear distinctions and cross-reference each other.

**lookup-join.md (syntax reference)**:
- add link to high-level landing page
- improved parameter formatting and descriptions
- fixed template variable from `{esql}` to `{{esql}}`

**esql-lookup-join.md (landing page)**:

https://github.com/elastic/elasticsearch/pull/127215/commits/9d23f244c7e200ff279ff5f2d245c9d5a53db50c adds improved example with setup code


improve layout:
- added "compare with enrich" section header
- simplified "how the command works" with clearer parameter explanation
- added code example in how it works section
- improved image alt text for accessibility
- organized example section with better context and SQL comparison
- added dropdown for sample tables to reduce visual clutter
- added "query" subheading for clearer organization
- included reference to additional examples in command reference

leemthompo committed:
[DOCS][ESQL] Cleanup and cross-reference LOOKUP JOIN reference and landing pages

**lookup-join.md (syntax reference)**:
- removed tip formatting for simpler direct link to landing page
- improved parameter formatting and descriptions
- fixed template variable from `{esql}` to `{{esql}}`

**esql-lookup-join.md (landing page)**:
- added "compare with enrich" section header
- simplified "how the command works" with clearer parameter explanation
- added code example in how it works section
- improved image alt text for accessibility
- organized example section with better context and SQL comparison
- added dropdown for sample tables to reduce visual clutter
- added "query" subheading for clearer organization
- included reference to additional examples in command reference
- removed excessive whitespace

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

alex-spies commented on docs/reference/query-languages/esql/esql-lookup-join.md:
```diff
@@ -26,25 +28,31 @@ For example, you can use `LOOKUP JOIN` to:
+### Sample tables
 
-`LOOKUP JOIN` has left-join behavior. If no rows match in the lookup index, `LOOKUP JOIN` retains the incoming row and adds `null`s. If many rows in the lookup index match, `LOOKUP JOIN` adds one row per match.
+In this example, we have two sample tables, `employees` and `languages_non_unique_key`. The `employees` table contains employee information, including their language code. The `languages_non_unique_key` table contains language codes and their corresponding names and countries.
```
out of scope: `languages_non_unique_key` is a weird name for an index IMO; it makes sense in the context of the tests that this is taken from, but doesn't mean much to readers of this page, I think.

This extends to the example query below, which doesn't provide a meaningful example from real life. It'd be nicer to have an example that is intuitive for new users of this command.

The same applies to some of the examples in the reference entry (in `lookup-join.md`) that were taken directly from our test. The threat hunting examples are much better, in contrast.

leemthompo replied:
I agree, I think we should just make the example on the landing page be super simple. I can do that in this PR. Then we can think about how to improve the examples in the reference entry in a follow up and/or think about more end-to-end tutorials.

alex-spies replied:
I didn't want to blow the scope of this PR, but if you'd like to add a better example right away, that's great! I'm happy either way :)

leemthompo replied:
I think we can work with what we have just make it a little clearer and user-friendly :)

leemthompo replied:
for folks following along: https://github.com/elastic/elasticsearch/pull/127215/commits/9d23f244c7e200ff279ff5f2d245c9d5a53db50c

alex-spies commented on docs/reference/query-languages/esql/esql-lookup-join.md:
```diff
@@ -26,25 +28,31 @@ For example, you can use `LOOKUP JOIN` to:
+In this example, we have two sample tables, `employees` and `languages_non_unique_key`. The `employees` table contains employee information, including their language code. The `languages_non_unique_key` table contains language codes and their corresponding names and countries.
 
-In this example, we have two sample tables:
+:::{dropdown} Show the sample tables
```
The dropdown is great and looks much cleaner!

alex-spies commented on docs/reference/query-languages/esql/esql-lookup-join.md:
```diff
@@ -26,25 +28,31 @@ For example, you can use `LOOKUP JOIN` to:
+### Sample tables
 
-`LOOKUP JOIN` has left-join behavior. If no rows match in the lookup index, `LOOKUP JOIN` retains the incoming row and adds `null`s. If many rows in the lookup index match, `LOOKUP JOIN` adds one row per match.
+In this example, we have two sample tables, `employees` and `languages_non_unique_key`. The `employees` table contains employee information, including their language code. The `languages_non_unique_key` table contains language codes and their corresponding names and countries.
```
Employees don't have a language code, we construct this using a modulus operation just to have some way to correlate the two tables. If we want to stick with this example, we can maybe reword like this:
```suggestion
In this example, we have two sample tables, `employees` and `languages_non_unique_key`.  The `languages_non_unique_key` table contains languages together with countries where the language is spoken and a code that uniquely identifies the language. The `employees` table contains employee information, including their employee numbers.
```

And below, under `Query`, we could write:
```
For the sake of the example, we construct some arbitrary code from employee numbers and use it to assign a language based on this code.
```

leemthompo replied:
Thanks @alex-spies, I'm thinking for the sake of this illustrative example we can just rename the table to `languages`?

alex-spies replied:
Oh, right! I thought this example was included via `:::{include} ../examples/docs-lookup-join.csv-spec...` but it's actually hand-written.

I guess we can rename the table for now; I wouldn't polish this too much, though, given that we want to use a nicer example, anyway.

(And, ideally, that example should live in our tests and be included rather than being hand-written again. I'm happy to help place the example into our tests once we have it :) )

leemthompo replied:
blown away by https://github.com/elastic/elasticsearch/pull/127215/commits/9d23f244c7e200ff279ff5f2d245c9d5a53db50c

alex-spies commented on docs/reference/query-languages/esql/esql-lookup-join.md:
```diff
@@ -100,9 +111,12 @@ FROM employees
 
+### Additional examples
+
+Refer to the examples section of the [`LOOKUP JOIN`](/reference/query-languages/esql/commands/processing-commands.md#esql-lookup-join) command reference for more examples.
```
++

alex-spies reviewed (approved):
Thanks Liam! I have some minor remarks and suggestions, but otherwise this LGTM and it's a nice improvement over the previous version.

Please feel free to address the remarks at your own discretion.

Also, we probably want to manually backport (an adapted version of) this to the 8.18/8.x docs, right?

leemthompo committed:
Improve example, add setup code

replaced abstract employee/language example with security monitoring use case
added setup instructions for creating test indices
included sample data loading via bulk api
new practical query example joining firewall logs with threat data
simplified results output showing threat detection scenario
added note about left-join behavior
improved code comments and structure
added required index.mode: lookup setting info

alex-spies commented on docs/reference/query-languages/esql/esql-lookup-join.md:
Much nicer example, thank you!

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/127232)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/127232">](https://github.com/elastic/elasticsearch/pull/127232) |
